twisted/conch/client/default.py

```diff
@@ -246,18 +255,46 @@ def getPrivateKey(self):
 
     def getGenericAnswers(self, name, instruction, prompts):
         responses = []
-        try:
-            oldout, oldin = sys.stdout, sys.stdin
-            sys.stdin = sys.stdout = open('/dev/tty','r+')
+        with self._replaceStdoutStdin():
             if name:
-                print(name)
+                print(name.decode("UTF8"))
             if instruction:
-                print(instruction)
+                print(instruction.decode("UTF8"))
             for prompt, echo in prompts:
+                prompt = prompt.decode("UTF8")
                 if echo:
                     responses.append(raw_input(prompt))
                 else:
                     responses.append(getpass.getpass(prompt))
-        finally:
-            sys.stdout,sys.stdin=oldout,oldin
         return defer.succeed(responses)
+
+
+    @classmethod
+    def _openTty(cls):
+        """
+        Open /dev/tty as two streams one in read, one in write mode,
+        and return them.
+        """
+        stdin = open("/dev/tty", "rb")
+        stdout = open("/dev/tty", "wb")
+        if _PY3:
+            stdin = io.TextIOWrapper(stdin)
+            stdout = io.TextIOWrapper(stdout)
+        return stdin, stdout
+
+
+    @classmethod
+    @contextlib.contextmanager
+    def _replaceStdoutStdin(cls):
```

Comment:
Another (optional) thought - maybe this should be somewhere more central?  there are a dozen other places in Twisted that set sys.stdout; some of them might benefit from the improved discipline of a contextmanager as well.